### PR TITLE
feat: `startBroadcast` and `stopBroadcast` cheatcodes

### PR DESCRIPTION
**Pull request by Karrq**
# What :computer: 
* Add support for `startBroadcast` and `stopBroadcast` cheatcodes
* Related tests

# Evidence :camera:
image

# Notes:
The broadcast tests are a bit of a false positive, the broadcast cheatcodes collect call data to generate transactions that can be broadcasted on chain - thus it has no observable effect in solidity.

### CODE REVIEW

**Comment by dutterbutter** [comment]
@MexicanAce this is necessary for truly testing `zkforge script` cmd fyi.

File: crates/era-cheatcodes/src/cheatcodes.rs [hunk 1]
```diff
@@ -321,6 +334,24 @@ impl<S: DatabaseExt + Send, H: HistoryMode> DynTracer<EraDb<S>, SimpleMemory<H>>
             }
         }
 
+        if let Opcode::Ret(_) = data.opcode.variant.opcode {
+            let current = state.vm_local_state.callstack.current;
+            if current.code_address != CHEATCODE_ADDRESS {
+                if let Some(broadcast) = &self.permanent_actions.broadcast {
+                    if state.vm_local_state.callstack.depth() == broadcast.depth
+                    // - 1
```
Comment (nbaztec) [inline crates/era-cheatcodes/src/cheatcodes.rs:342]:
```suggestion
```

File: crates/era-cheatcodes/src/cheatcodes.rs [hunk 2]
```diff
@@ -330,6 +361,76 @@ impl<S: DatabaseExt + Send, H: HistoryMode> DynTracer<EraDb<S>, SimpleMemory<H>>
...
+                                        value: Some(farcall_abi.ergs_passed.into()),
+                                        data: Some(get_calldata(&state, memory).into()),
+                                        nonce: Some(nonce),
+                                        ..Default::default()
+                                    },
+                                ),
+                        };
+                        tracing::debug!(?tx, "storing for broadcast");
+
+                        self.broadcastable_transactions.push(tx);
```
Comment (nbaztec) [inline crates/era-cheatcodes/src/cheatcodes.rs:425]: Are we using this variable somehow, or is it for a later PR?

Comment (Karrq) [reply crates/era-cheatcodes/src/cheatcodes.rs:425]: `tx` no, it's just for debugging, meanwhile `broadcastable_transactions` is the actual output of the cheatcode

Comment (nbaztec) [reply crates/era-cheatcodes/src/cheatcodes.rs:425]: Wondering how these transactions would be applied

File: crates/era-cheatcodes/src/cheatcodes.rs [hunk 3]
```diff
@@ -1402,6 +1566,88 @@ impl CheatcodeTracer {
...
+        new_origin: Option<H160>,
+    ) {
+        if self.permanent_actions.start_prank.is_some() {
+            tracing::error!("broadcast is incompatible with prank");
+            return
+        }
+
+        let depth = state.vm_local_state.callstack.depth() // - 2
+            ;
+
```
Comment (nbaztec) [inline crates/era-cheatcodes/src/cheatcodes.rs:1626]:
```suggestion
        let depth = state.vm_local_state.callstack.depth();

```

**Review by nbaztec — APPROVED** [review]

File: crates/era-cheatcodes/src/cheatcodes.rs (hunk 2 above)
Comment (Karrq) [reply crates/era-cheatcodes/src/cheatcodes.rs:425]: This information is used in scripting

**Comment by Karrq** [comment]
Rebased on top of `main`

**Pull request merged**